pkg/maps/configmap/config_map.go

```diff
@@ -0,0 +1,181 @@
+	// of the same size (64 bytes each).
+	MaxEntries = 256
+)
+
+// ConfigKey is the index to the Unix time offset. Only value 0 is supported for now.
+// +k8s:deepcopy-gen=true
+// +k8s:deepcopy-gen:interfaces=github.com/cilium/cilium/pkg/bpf.MapKey
+type ConfigKey struct {
+	key uint32 `align:"ctx"`
+}
+
+// ConfigValue is the generic datapath runtime config value.
+// +k8s:deepcopy-gen=true
+// +k8s:deepcopy-gen:interfaces=github.com/cilium/cilium/pkg/bpf.MapValue
+type ConfigValue struct {
+	reserved [8]uint64
+}
+
+// ConfigValue0 is the first 64 bytes of datapath runtime config.
+// +k8s:deepcopy-gen=true
+// +k8s:deepcopy-gen:interfaces=github.com/cilium/cilium/pkg/bpf.MapValue
+type ConfigValue0 struct {
+	reserved [8]uint64 `align:"reserved"`
+}
+
+// ConfigValues is the set of all config values for the local cache
+type ConfigValues struct {
+	value0 ConfigValue0
+}
+
+// String pretty print the ConfigKey
+func (k *ConfigKey) String() string {
+	return fmt.Sprintf("%d", k.key)
+}
+
+// String pretty print the ConfigValues.
+func (v *ConfigValues) String() string {
+	return v.value0.String()
+}
+
+// String pretty print the ConfigValue.
+func (v *ConfigValue) String() string {
+	return fmt.Sprintf("<NOT TO BE USED DIRECTLY>")
+}
+
+// String pretty print the ConfigValue0.
+func (v *ConfigValue0) String() string {
+	return ""
+}
+
+// GetValuePtr returns the unsafe pointer to the BPF value.
+func (v *ConfigValue) GetValuePtr() unsafe.Pointer { return unsafe.Pointer(v) }
+
+// GetValuePtr returns the unsafe pointer to the BPF value.
+func (v *ConfigValue0) GetValuePtr() unsafe.Pointer { return unsafe.Pointer(v) }
+
+// GetKeyPtr returns the unsafe pointer to the BPF key
+func (k *ConfigKey) GetKeyPtr() unsafe.Pointer { return unsafe.Pointer(k) }
+
+// NewValue returns a new empty instance of the structure represeting the BPF
+// map value
+func (k *ConfigKey) NewValue() bpf.MapValue { return &ConfigValue{} }
+
+func newConfigKey(key uint32) *ConfigKey {
+	return &ConfigKey{
+		key: key,
+	}
+}
+
+type ConfigMap struct {
+	*bpf.Map
+	config ConfigValues // current config
+}
+
+var (
+	once      sync.Once
+	configMap ConfigMap
+)
+
+type ControllerManager interface {
+	UpdateController(name string, params controller.ControllerParams)
+}
+
+// MapCreate will create an config map
+func InitMap(ctx context.Context, controllerManager ControllerManager) error {
+	once.Do(func() {
+		configMap = ConfigMap{
+			Map: bpf.NewMap(MapName,
+				bpf.MapTypeArray,
+				&ConfigKey{},
+				int(unsafe.Sizeof(ConfigKey{})),
+				&ConfigValue{},
+				int(unsafe.Sizeof(ConfigValue{})),
+				MaxEntries,
+				0, 0,
+				bpf.ConvertKeyValue,
+			),
+		}
+
+		configMap.syncConfig(true)
+
+		// Add controller for keeping clock in sync for NTP time jumps and any difference
+		// between monotonic and boottime clocks.
+		ctrlName := "sync-config"
+		controllerManager.UpdateController(ctrlName,
```

Comment:
Newbie question: how come that the configmap sets up the periodic task? Is that a convention? I kind of expected utime or similar to do that. Thinking this forward, we'd end up with a bit of custom code here for each config, which doesn't scale well.

Reply:
That is a design question rather than a newbie one. I had no other use than `utime` for this in mind, so went with a minima interface between `configmap` and `utime`. Open for better proposals :-)